crates/shared/src/sources/balancer_v2/graph_api.rs

```diff
@@ -28,29 +33,39 @@ const QUERY_PAGE_SIZE: usize = 10;
 ///
 /// This client is not implemented to allow general GraphQL queries, but instead
 /// implements high-level methods that perform GraphQL queries under the hood.
-pub struct BalancerSubgraphClient(SubgraphClient);
+pub struct BalancerSubgraphClient {
+    pub graph: SubgraphClient,
+    pub web3: Web3,
+}
 
 impl BalancerSubgraphClient {
     /// Creates a new Balancer subgraph client for the specified chain ID.
-    pub fn for_chain(chain_id: u64, client: Client) -> Result<Self> {
+    pub fn for_chain(chain_id: u64, client: Client, web3: Web3) -> Result<Self> {
         let subgraph_name = match chain_id {
             1 => "balancer-v2",
             4 => "balancer-rinkeby-v2",
             5 => "balancer-goerli-v2",
             _ => bail!("unsupported chain {}", chain_id),
         };
-        Ok(Self(SubgraphClient::new(
-            "balancer-labs",
-            subgraph_name,
-            client,
-        )?))
+        Ok(BalancerSubgraphClient {
+            graph: SubgraphClient::new("balancer-labs", subgraph_name, client)?,
+            web3,
+        })
     }
 
     /// Retrieves the list of registered pools from the subgraph.
     pub async fn get_registered_pools(&self) -> Result<RegisteredPools> {
         use self::pools_query::*;
 
         let block_number = self.get_safe_block().await?;
+        let block_hash = self
+            .web3
+            .eth()
+            .block(U64::from(block_number).into())
+            .await?
+            .context("missing block")?
+            .hash
+            .context("no hash in block - pending block")?;
```

Comment:
I don't think the `Subgraph` client should be responsible for this - specifically, it should only fetch the subgraph data.

Reply:
Fixed with the change below.